pkg/trace/sampling.go

```diff
@@ -0,0 +1,285 @@
+// "SampleRate" tag that Honeycomb can use for sample rate adjustment.
+//
+// This is a hack.  We hope that one day we can move to an official
+// OpenTelemetry head-based sampling mechanism that handles this more
+// cleanly. [1] At the time of writing, this part of the standard is in flux and
+// not all OpenTelemetry libraries support it.
+//
+// [1] https://opentelemetry.io/docs/specs/otel/trace/tracestate-probability-sampling/
+func NewSampleRateTaggingSampler(defaultSampleRate uint) sdktrace.Sampler {
+	return &sampleRateTaggingSampler{
+		defaultSampleRate,
+	}
+}
+
+// Description returns a description of this `sampleRateTaggingSampler`.
+func (s *sampleRateTaggingSampler) Description() string {
+	return "sampleRateTaggingSampler"
+}
+
+// ShouldSample returns "record and sample" if the parent span context is sampled
+// and "drop" otherwise.
+//
+// # In the case where the parent span context is sampled, we also attach a sample
+//
+// rate attribute to this span.  This hack is the main purrpose of this sampler.
+//
+//nolint:gocritic // Why: required by otel
+func (s *sampleRateTaggingSampler) ShouldSample(p sdktrace.SamplingParameters) sdktrace.SamplingResult {
+	psc := trace.SpanContextFromContext(p.ParentContext)
+	ts := psc.TraceState()
+
+	if psc.IsSampled() {
+		sampleRate, ok := GetTraceStateSampleRate(ts)
+		if ok {
+			return sdktrace.SamplingResult{
+				Decision:   sdktrace.RecordAndSample,
+				Tracestate: ts,
+				Attributes: []attribute.KeyValue{
+					sampleRateAttribute.Int(int(sampleRate)),
+				},
+			}
+		}
+
+		// Sampled but no "sample_rate" in the tracestate.
+		return sdktrace.SamplingResult{
+			Decision:   sdktrace.RecordAndSample,
+			Tracestate: ts,
+			Attributes: []attribute.KeyValue{
+				sampleRateAttribute.Int(int(s.defaultSampleRate)),
+			},
+		}
+	}
+
+	// Not sampled.
+	return sdktrace.SamplingResult{
+		Decision:   sdktrace.Drop,
+		Tracestate: ts,
+	}
+}
+
+// sampleRateKey is the key we use to store the sample rate in the tracestate.
+//
+// This value goes into the tracecontext header [1] that gets sent to downstream
+// services.  This could be used to forward sample rate information to other
+// services, allowing them to populate their sample rate tags without guessing.
+//
+// We don't do this currently.  We can get away with guessing for now because
+// our sample rates tend to line up across all services.  Hopefully we'll be
+// able to move to an official standard before this becomes necessary.
+//
+// [1] https://www.w3.org/TR/trace-context/
+const sampleRateKey = "sample_rate"
+
+// SetTraceStateSampleRate sets the sample rate in the tracestate.
+func SetTraceStateSampleRate(ts trace.TraceState, sampleRate uint) trace.TraceState {
+	updated, err := ts.Insert(sampleRateKey, fmt.Sprintf("%d", sampleRate))
+	if err != nil {
+		// This really shouldn't happen.
+		log.Warn(context.TODO(), "failed to insert sample rate into tracestate", events.NewErrorInfo(err))
```

Comment:
nit: This is a huge nit, but I reserve `context.TODO()` for cases where we plan on updating the signature to include a context. This doesn't seem like something we would go add it to (if we did, we would do it now, right?). So, I would say we should use `context.Background` instead.